src/main/java/org/threadly/litesockets/SocketExecuterCommonBase.java

```diff
@@ -56,73 +57,90 @@
     }
   }
-
+  
   @Override
-  public void removeClient(Client client) {
-    ArgumentVerifier.assertNotNull(client, "Client");
-    if(isRunning() && clients.remove(client.getChannel()) != null) {
-      readScheduler.execute(new RemoveFromSelector(client.getChannel(), readSelector));
-      writeScheduler.execute(new RemoveFromSelector(client.getChannel(), writeSelector));
-      readSelector.wakeup();
-      writeSelector.wakeup();
+  public TCPClient createTCPClient(final String host, final int port) throws IOException {
+    checkRunning();
+    return new TCPClient(this, host, port);
+  }
+  
+  
+  @Override
+  public TCPClient createTCPClient(final SocketChannel sc) throws IOException {
+    checkRunning();
+    final TCPClient tc = new TCPClient(this, sc);
+    this.setClientOperations(tc);
+    return tc;
+  }
+  
+  @Override
+  public TCPServer createTCPServer(final String host, final int port) throws IOException {
+    checkRunning();
+    return new TCPServer(this, host, port);
+  }
+  
+  @Override
+  public TCPServer createTCPServer(final ServerSocketChannel ssc) throws IOException {
+    checkRunning();
+    return new TCPServer(this, ssc);
+  }
+  
+  @Override
+  public UDPServer createUDPServer(final String host, final int port) throws IOException {
+    checkRunning();
+    return new UDPServer(this, host, port);
+  }
+  
+  private boolean checkAndAddServer(final Server server) {
+    try{
+      checkRunning();
+    } catch(Exception e) {
+      return false;
+    }
+    if(server.isClosed() || server.getSocketExecuter() != this) {
+      servers.remove(server);
+      return false;
+    }else if(!servers.containsKey(server.getSelectableChannel())) {
+      servers.putIfAbsent(server.getSelectableChannel(), server);
     }
+    return true;
   }
-
+  
   @Override
-  public void addServer(final Server server) {
-    ArgumentVerifier.assertNotNull(server, "Server");
-    if(isRunning()) {
-      Server sn = servers.putIfAbsent(server.getSelectableChannel(), server);
-      if(sn == null) {
-        server.setSocketExecuter(this);
-        server.setThreadExecutor(schedulerPool);
-        if(server.getServerType() == WireProtocol.TCP) {
-          acceptScheduler.execute(new AddToSelector(server, acceptSelector, SelectionKey.OP_ACCEPT));
-        } else if(server.getServerType() == WireProtocol.UDP) {
-          acceptScheduler.execute(new AddToSelector(server, acceptSelector, SelectionKey.OP_READ));
-        }
-        acceptSelector.wakeup();
+  public void startListening(final Server server) {
+    if(!checkAndAddServer(server)) {
+      return;
+    } else {
+      if(server.getServerType() == WireProtocol.TCP) {
+        acceptScheduler.execute(new AddToSelector(server, acceptSelector, SelectionKey.OP_ACCEPT));
+      } else if(server.getServerType() == WireProtocol.UDP) {
+        acceptScheduler.execute(new AddToSelector(server, acceptSelector, SelectionKey.OP_READ));
       }
```

Comment:
Does it make sense to add an else condition as a safety check (throw UnsupportedOperationException)...same for the one below